Render CustomSlurmSettings in cookbook

jdeamicis opened the pull request:
### Description of changes
* Render CustomSlurmSettings in cookbook
  * create include file infrastructure for user-controlled custom Slurm settings defined in the cluster config YAML;
  * add possibility to update the bulk `CustomSlurmSettings` in the cookbook;
  * create unit test for the renderization of the bulk custom Slurm settings;

### Tests
* Ran manual tests with:
  * created one cluster with `CustomSlurmSettings`; 
  * started with cluster with no CustomSlurmSettings (to validate that normal cluster creation is possible without the added functionality) and then:
    * updated the cluster with `CustomSlurmSettings`;
    * updated the cluster modifying `CustomSlurmSettings`;
    * updated the cluster removing `ClusterSlurmSettings`;
* Added unit tests to cover the renderization of the custom Slurm settings from the `CustomSlurmSettings` coming from the YAML file.
* Ran round of selected integration tests with the development branch used here.

### References
* PR from the CLI side: https://github.com/aws/aws-parallelcluster/pull/5103

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [ ] ~Check if documentation is impacted by this change.~ -> Pending later

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

jdeamicis committed:
Render CustomSlurmSettings in cookbook

Create script to generate custom Slurm settings include files.

Add include file under slurm.conf if `CustomSlurmSettings` is used.

Add unit test for the generation of custom Slurm settings include
files.

Small modifications to improve readability

Add header to test_slurm_config_generator.py.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Allow generation of empty custom Slurm setting include file

Always generate a custom Slurm setting include file, which will be
empty if no custom Slurm settings are used.

This avoids additional logic to delete the custom Slurm settings
include files if the custom Slurm settings are removed via a pcluster
update-cluster operation.

Add unit test to verify the generation of an empty include file if
CustomSlurmSettings is not used.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Rebuild custom Slurm settings include files at cluster update

Allow rebuilding the custom Slurm settings includes in update_head_node.rb
if CustomSlurmSettings have changed.

Perform a restart of clustermgtd, restart of slurmctld and Slurm reconfig
when custom Slurm settings are changed.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

NSsirena commented on libraries/helpers.rb:
```diff
@@ -310,6 +310,14 @@ def are_queues_updated?
   config["Scheduling"] != previous_config["Scheduling"] or is_compute_node_bootstrap_timeout_updated?(previous_config, config)
 end
 
+# Verify if CustomSlurmSettings has been updated in the config
```
We should have tests for these functions.
They should be added as part of the review of the recipes, but let's try to add them opportunistically.

Let's open a nice-to-have task to add some of them before closing the story.

jdeamicis replied:
I think this is part of some broader story the QA team is working on.

NSsirena commented on libraries/helpers.rb:
```diff
@@ -310,6 +310,14 @@ def are_queues_updated?
 end
 
+# Verify if CustomSlurmSettings has been updated in the config
+def are_custom_slurm_settings_updated?
```
Since this function is monitoring only the `bulk` custom settings, let's be explicit about this and rename the function `are_bulk_custom_slurm_settings_updated?`

jdeamicis replied:
Addressing.

jdeamicis replied:
Done.

NSsirena commented on cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/pcluster_custom_slurm_settings_include_file_generator.py:
```diff
@@ -0,0 +1,169 @@
+# Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
2023 ?

jdeamicis replied:
Addressing in next commit.

jdeamicis replied:
Done.

NSsirena commented on cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/pcluster_custom_slurm_settings_include_file_generator.py:
```diff
@@ -0,0 +1,169 @@
+    :param param: dictionary containing the parameter structure.
+    :return: string to be included in the Slurm configuration file.
+    """
+    key = list(param.keys())[0]
+    value = list(param.values())[0]
```
What if the dictionary has more than one item ?

We should at least check this condition and add a log line to track it.

jdeamicis replied:
You mean having an explicit assertion during execution?

jdeamicis replied:
Removed this method.

NSsirena commented on cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/pcluster_custom_slurm_settings_include_file_generator.py:
```diff
@@ -0,0 +1,169 @@
+            first_subparam = {key: value}
+            break
+
+    remaining_subparams = {key: value for key, value in param.items() if key not in first_subparam}
```
With the current code this will fail if the dictionary doesn't contains at least one of the keys listed at line `66` since `first_subparam` will not be initialized.

jdeamicis replied:
Addressing in the next commit.

jdeamicis replied:
Done.

NSsirena commented on cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/pcluster_custom_slurm_settings_include_file_generator.py:
```diff
@@ -0,0 +1,169 @@
+    # (e.g. the NodeName parameter in the node definition)
+    for key, value in param.items():
+        # Here we assume that the users do not pass more than one of these parameters in the same dictionary.
+        if key.lower() in ["nodename", "partitionname", "nodeset", "downnodes"]:
```
We should check that no more than one of these keys is present or fail.

Also, unless you want to make this function more general and handle both cases you should also check that at least one is present.

On the other and, if you initialize `first_subparam` with `None` and then skip the first `for` loop below if `first_subparam is None` you should be able to handle both cases with this function and remove the previous function.

jdeamicis replied:
> We should check that no more than one of these keys is present or fail.

I'm not sure I agree with that.

I mean, I see the problem, but that would be a sort of validation performed at cookbook time: we should either do it in the CLI at validator time, or not do it at all since a main assumption in the feature is that we are not going to validate the Slurm config coming from the user, other than checking against the deny-list. Let's discuss about this.

> Also, unless you want to make this function more general and handle both cases you should also check that at least one is present.

Yes, I'm going to handle both cases with the same method. Plus I wonder if there are other configuration parameters that may be composed of space-separated key=value pairs that are not covered by that list...

jdeamicis replied:
OK, we will put a warning if more than one of the special parameters in the list are provided.

jdeamicis replied:
Addressed.

NSsirena commented on cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/pcluster_custom_slurm_settings_include_file_generator.py:
```diff
@@ -0,0 +1,169 @@
+    )
+
+
+def _generate_simple_parameter(param: Dict) -> str:
```
While I understand from where it comes from, I really do not like this method. 
A method that assumes a dictionary has only a single key/value pair is confusing.

You should be able to handle this case within the boundaries of the following method when
* no `first_subparam` is passed
* only a key/value pair is present.

jdeamicis replied:
Removed method.

NSsirena commented on cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/pcluster_custom_slurm_settings_include_file_generator.py:
```diff
@@ -0,0 +1,169 @@
+    return f"{key}={value}"
+
+
+def _generate_complex_parameter(param: Dict) -> str:
```
probably a generic `_render_parameter` would be better (provided that both cases are handled with the same function.

jdeamicis replied:
OK, I see what you mean. Addressing in the next commit.

jdeamicis replied:
Done.

NSsirena commented on cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/pcluster_custom_slurm_settings_include_file_generator.py:
```diff
@@ -0,0 +1,169 @@
+    # Load cluster configuration from YAML
+    cluster_config = _load_cluster_config(input_file)
+
+    for include_target in ["slurm"]:
```
Why this generalization ?

If we do not have an immediate need I'd keep it simple and use 

```
include_file_config = cluster_config["Scheduling"]["SlurmSettings"].get("CustomSlurmSettings", {} )
```

and also remove the `include_target` parameter.

Also since these files are `slurm` specific, do we expect to have other targets in the future ?

jdeamicis replied:
I was thinking about future extensions where we allow setting `CustomSlurmDbdSettings`, `CustomSlurmCgroupSettings`, etc.

If we don't want to do this now I can remove it. Let's discuss about this.

jdeamicis replied:
We decided to remove it.

jdeamicis replied:
Addressed.

NSsirena commented on cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/pcluster_custom_slurm_settings_include_file_generator.py:
```diff
@@ -0,0 +1,169 @@
+                dryrun,
+            )
+
+    log.info("Finished.")
```
As is, it is not very informative :).

Let's add a more descriptive message like `Definition of Custom Slurm Settings completed.`

jdeamicis replied:
It's inherited from the script I used as a blueprint for this one :)

Consider that this would appear after the following line:
```python
log.info("Generating custom_slurm_settings_include_file_%s.conf", include_target)
```

jdeamicis replied:
Addressed.

jdeamicis committed:
Address PR review comments

Use one single method to render any parameter for the CustomSlurmSettings.

Small refactoring (changed some funtions and method names).

Remove unnecessary generalization to multiple files for custom Slurm
settings include file generation.

Add more logging of the internals of the code.

Add one parameter to unit test for rendering of include file.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

NSsirena commented on cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/pcluster_custom_slurm_settings_include_file_generator.py:
```diff
@@ -0,0 +1,159 @@
+    for key, value in param.items():
+        # Here we assume that the users do not pass more than one of these parameters in the same dictionary.
+        if key.lower() in ["nodename", "partitionname", "nodeset", "downnodes"]:
+            if not first_subparam_found:
```
[Nitpick] Here we could write 

`if !first_subparam:` and avoid using a bool flag.

jdeamicis replied:
Done.
